### PR TITLE
feat(ui-shell): adds UX patterns for Header with SideNav

### LINKED ISSUE #2801: Implement new UX Patterns for Side Nav
This includes: 

- [x] Fixed SideNav with open / closing via hamburger menu
- [x] SideNav with rail + hover interactions

### PR DESCRIPTION
**Pull request by jnm2377**
Closes #2801

Adds UX for `Header` with responsive `SideNav`. `SideNav` will be visible until `lg` breakpoint, then it will become toggle-able with the `HeaderMenuButton`.

#### Changelog

**New**

- adds expanded state for `Header`
- props for `Header` and `SideNav`
- scss selector `.#{$prefix}--side-nav--ux` for `SideNav` (which adds styles for displaying until a certain browser width, since it then becomes toggle-able)
- scss selector `.#{$prefix}--header__menu-toggle` for `HeaderMenuButton` (doesn't display it until certain browser width when the SideNav becomes responsive)

### CODE REVIEW

File: packages/components/src/components/ui-shell/_side-nav.scss [hunk 1]
```diff
@@ -157,6 +170,11 @@
     @include expanded($opacity: true, $visibility: true);
   }
 
+  .#{$prefix}--side-nav--ux .#{$prefix}--side-nav__details {
+    opacity: 1;
+    visibility: visible;
```
Comment (asudoh) [inline packages/components/src/components/ui-shell/_side-nav.scss]: Due to [the problem with `visibility: visible`](https://jakearchibald.com/2014/visible-undoes-hidden/):

```suggestion
    visibility: inherit;
```

File: packages/react/src/components/UIShell/Header.js [hunk 2]
```diff
@@ -8,16 +8,41 @@
...
+
+  const childrenProps = {
+    function: handleHeaderMenuButtonClick,
+    state: isSideNavExpandedState,
+  };
+
   return (
     <header {...rest} className={className} role="banner">
-      {children}
+      <HandleMenuButtonClickContext.Provider value={childrenProps}>
```
Comment (asudoh) [inline packages/react/src/components/UIShell/Header.js]: Is context usage here is for [the same problem discussed in #2786](https://github.com/carbon-design-system/carbon/pull/2786#issuecomment-494174254)? If so, my recommendation is creating a "state manager" like:

```javascript
const HeaderContainer = ({
  defaultIsSideNavExpanded,
  isSideNavExpanded,
  onToggleSideNav,
  render: Children,
}) => {
  const [isSideNavExpandedState, setIsSideNavExpandedState] = useState(
    defaultIsSideNavExpanded
  );
  const handleToggleSideNav = useCallback(
    (
      evt,
      {
        isExpanded = typeof isSideNavExpanded === 'undefined'
          ? !isSideNavExpandedState
          : !isSideNavExpanded,
      } = {}
    ) => {
      if (typeof isSideNavExpanded === 'undefined') {
        setIsSideNavExpandedState(isExpanded);
      }
      if (onToggleSideNav) {
        onToggleSideNav(evt, { isSideNavExpanded: isExpanded });
      }
    },
    [
      isSideNavExpandedState,
      setIsSideNavExpandedState,
      isSideNavExpanded,
      onToggleSideNav,
    ]
  );
  return (
    <Children
      isSideNavExpanded={
        typeof isSideNavExpanded !== 'undefined'
          ? isSideNavExpanded
          : isSideNavExpandedState
      }
      onToggleSideNav={handleToggleSideNav}
    />
  );
};
```

Which can be used like:

```javascript
<HeaderContainer render={({ isSideNavExpanded, onClickSideNavExpando }) => (
  <>
    <Header aria-label="IBM Platform Name">
      <SkipToContent />
      <HeaderMenuButton
        aria-label="Open menu"
        isActive={isSideNavExpanded}
        onClick={onClickSideNavExpando}
      />
      ...
    </Header>
    <SideNav isSideRail aria-label="Side navigation" isExpanded={isSideNavExpanded}>
        ...
    </SideNav>
    <StoryContent />
  </>
)}>
```

**Review by asudoh — COMMENTED** [review]
Great to see progress here @jnm2377!

File: packages/react/src/components/UIShell/Header.js (hunk 2 above)
Comment (tw15egan) [reply packages/react/src/components/UIShell/Header.js]: Yes, this is the same issue as that comment. State manager like ☝️ may make maintainability a bit easier in the future.

**Review by asudoh — APPROVED** [review]
🎉 - Thanks @jnm2377!

**Review by tw15egan — APPROVED** [review]
🎉  Looks great!

**Pull request merged**